src/wrapper.rs

```diff
@@ -36,16 +37,40 @@ macro_rules! check_hr_wrapped {
     }};
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct DxcBlob {
-    inner: ComPtr<IDxcBlob>,
+    pub(crate) inner: ComPtr<IDxcBlob>,
 }
 
 impl DxcBlob {
-    fn new(inner: ComPtr<IDxcBlob>) -> Self {
+    pub(crate) fn new(inner: ComPtr<IDxcBlob>) -> Self {
+        unsafe {
+            inner.add_ref();
+        }
         Self { inner }
     }
 
+    pub fn from_slice(slice: &[u8]) -> Result<Self, HassleError> {
+        let mut blob = ComPtr::<IDxcBlob>::new();
+
+        // This weird pointer cast from IDxcBlob to ID3DBlob is safe:
+        // See https://github.com/microsoft/DirectXShaderCompiler/blob/285490211f2cbb44444a23d5f1c329881b96a43b/docs/HLSLChanges.rst
+        // `The HLSL compiler avoids pulling in DirectX headers and defines an
+        // IDxcBlob interface that has the same layout and interface identifier (IID).`
+        let blob_ptr: *mut *mut IDxcBlob = blob.as_mut_ptr::<IDxcBlob>();
+        check_hr_wrapped!(
+            unsafe { winapi::um::d3dcompiler::D3DCreateBlob(slice.len(), blob_ptr as *mut *mut _) },
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    slice.as_ptr(),
+                    blob.get_buffer_pointer() as *mut u8,
+                    slice.len(),
+                );
```

Comment:
```suggestion
                let target = std::slice::from_raw_parts_mut(
                    blob.get_buffer_pointer() as *mut _,
                    slice.len(),
                );
                target.copy_from_slice(slice);
```

More idiomatic might be to provide a `.as(_mut)_slice()` function as well that returns `self` as `slice::from_raw_parts(_mut)`, see `fn to_vec` below for that.